lib/Core/BuildEngine.cpp

```diff
@@ -687,8 +740,13 @@ class BuildEngineImpl : public BuildDBDelegate {
     auto scanRecord = inputRuleInfo.getPendingScanRecord();
 
     // Wake up all of the pending scan requests.
-    for (const auto& request: scanRecord->deferredScanRequests) {
-      ruleInfosToScan.push_back(request);
+    for (auto& request: scanRecord->deferredScanRequests) {
+      if ((--request->numPending) == 0 && !request->isInScanQueue) {
```

Comment:
🤔 This means that we are going to wait for _all_ pending scan requests in the current batch before we process the deferred scan again.   Can we short circuit that for cases where the current scan has been determined as needs to run?

Reply:
If we short circuit we cannot use a simple numPending to keep track of when all are finished as we will be reprocessing all pending requests and some old might still be deferred. 

I committed a simpler version that just re-processes after each result is available. This has the drawback that we will end up with the same request in the deferredScanRequests several times as well as reprocessing unnecessarily.

To fix this I added another commit that uses an unordered_set instead. This has the drawback that it will make the order of deferredScanRequests random. Is this ok, or should I use both a vector and unordered_set?

Finally I added back so we have the same logic but with the short circuit. This with an unordered_set instead of a counter to keep track of when to requeue.

I don't have access to a benchmark so I don't know which of these is the best performing.

Reply:
I agree that always reprocessing seems suboptimal.  Indeed, it is not clear if the random ordering of deferredScanRequests would be problematic.  I think where you have landed with the combination of unordered set  and the short circuit seems like a good balance.